src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.JsonTranscoding/Internal/Json/MessageTypeInfoResolver.cs

```diff
@@ -0,0 +1,165 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Diagnostics.CodeAnalysis;
+using System.Text.Json;
+using System.Text.Json.Serialization.Metadata;
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using Grpc.Shared;
+using Type = System.Type;
+
+namespace Microsoft.AspNetCore.Grpc.JsonTranscoding.Internal.Json;
+
+internal sealed class MessageTypeInfoResolver : IJsonTypeInfoResolver
+{
+    private readonly JsonContext _context;
+
+    public MessageTypeInfoResolver(JsonContext context)
+    {
+        _context = context;
+    }
+
+    public JsonTypeInfo? GetTypeInfo(Type type, JsonSerializerOptions options)
+    {
+        if (!IsStandardMessage(type, out var messageDescriptor))
+        {
+            return null;
+        }
+
+        var typeInfo = JsonTypeInfo.CreateJsonTypeInfo(type, options);
+        typeInfo.CreateObject = () => Activator.CreateInstance(type)!;
+
+        var fields = messageDescriptor.Fields.InFieldNumberOrder();
+        var mappings = CreateJsonFieldMap(fields);
+
+        foreach (var field in fields)
+        {
+            mappings.Remove(field.JsonName);
```

Comment:
Why do you need add an entry to the dictionary with the `JsonName` if that will be discarded here?

Reply:
I've expanded the comments